### PR TITLE
More improvements to `ServerEnvironment` configuration

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**
This PR continues addressing #1315.

Enhancements introduced:
  * It is now possible to configure `ServerEnvironment ` features using `useSomething()` methods that accept a `Function<Class<? extends EnvironmentType>, Something>`.
 * In order to make the declaration a bit more compact `ServerEnvironment.Fn` functional interface was introduced. It's just a function, from the previous point, but with first generic argument hidden.
 * Tests of `ServerEnvironment` that related to the configuration were extracted into a separate test suite.

### CODE REVIEW

**Comment by dmitrykuzmin** [comment]
I don't know if PR in `base` is actually planned in scope of these changes, but I'd also propose to `@Internal`-ize the `Environment.register(...)` and maybe clarify `EnvironmentType` doc a bit, because from my experience it's not immediately apparent what exactly to do with a custom `EnvironmentType` once you've created it.

For example, with the current implementation, it's possible to employ `Environment.register(EnvironmentType)` to register a custom env type which actually accepts some dependencies on creation which help to deduce whether it's enabled or not. And then it would be logical to check for such env presence with `when(...)`, but this would throw an exception.

**Comment by alexander-yevsyukov** [comment]
@dmitrykuzmin, that would go into another PR, because `Environment` and `EnvironmentType` are in `base`. I'll add your notes to the governing issue.

File: server/src/main/java/io/spine/server/ServerEnvironment.java [hunk 1]
```diff
@@ -393,14 +400,20 @@ public StorageFactory storageFactory() {
...
-                        "Transport factory is not assigned for the current environment `%s`."
-                                + " Please call `.when(environmentType).use(transportFactory)`.",
-                        type.getSimpleName()));
+                .orElseThrow(() -> suggestConfiguringTransportFactory(type));
 
         return result;
     }
 
+    private static IllegalStateException
+    suggestConfiguringTransportFactory(Class<? extends EnvironmentType> type) {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/ServerEnvironment.java:409]: Should we have the same methods for the tracer factory and delivery?

If so, maybe we should arrange a private nested class `Suggest` with the names like `configuringTransportFactory`?

**Review by armiol — APPROVED** [review]
@alexander-yevsyukov LGTM with a suggestion to consider.

File: server/src/main/java/io/spine/server/ServerEnvironment.java (hunk 1 above)
Comment (armiol) [reply server/src/main/java/io/spine/server/ServerEnvironment.java:409]: Or event `SuggestConfiguring.transportFactory(..)`?

**Comment by alexander-yevsyukov** [comment]
@dmitrykuzmin 
> And then it would be logical to check for such env presence with when(...), but this would throw an exception.

`TypeConfigurator` now does this automatically in its constructor. I am planning to address improvements you suggest shortly in the PRs following in `base` and `core-java`.

**Pull request merged**